### PR TITLE
[SYCL] Fix warnings in source/detail code

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
Fixes places with copy instead of usage by reference or move.
Protected potential but unlikely constant overflow with exception.

### CODE REVIEW

File: sycl/source/detail/persistent_device_code_cache.cpp [hunk 1]
```diff
@@ -320,7 +320,7 @@ std::vector<std::vector<char>> PersistentDeviceCodeCache::getItemFromDisc(
 std::vector<std::vector<char>>
 PersistentDeviceCodeCache::getCompiledKernelFromDisc(
     const std::vector<device> &Devices, const std::string &BuildOptionsString,
-    const std::string SourceStr) {
+    const std::string &SourceStr) {
```
Comment (aelovikov-intel) [inline sycl/source/detail/persistent_device_code_cache.cpp:323]: nit: I think more idiomatic fix would be to use `std::string_view`.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 2]
```diff
@@ -990,27 +1008,23 @@ ur_program_handle_t ProgramManager::getBuiltURProgram(
     // emplace all subsets of the current set of devices into the cache.
     // Set of all devices is not included in the loop as it was already added
```
Comment (aelovikov-intel) [inline sycl/source/detail/program_manager/program_manager.cpp:1009]: I think this comment ("Set of all devices is not included...") needs to be moved to the line 832, if I understand the logic correctly.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 3]
```diff
@@ -990,27 +1008,23 @@ ur_program_handle_t ProgramManager::getBuiltURProgram(
...
     // into the cache.
-    for (int Mask = 1; Mask < (1 << URDevicesSet.size()) - 1; ++Mask) {
```
Comment (aelovikov-intel) [inline sycl/source/detail/program_manager/program_manager.cpp:993]: This code is much more maintainable than the new one. Also, if the problem is the "overflow" here, then it's inherent part of the problem. We'll have that many entries in the resulting set anyway (and likely OOM regardless).

Comment (KseniyaTikhomirova) [reply sycl/source/detail/program_manager/program_manager.cpp:993]: discussed in person, fixed in https://github.com/intel/llvm/pull/16334/commits/19983fe9bd65e3f9008cd3e244886c89dd547a3d

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 2 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/program_manager/program_manager.cpp:991]: fixed https://github.com/intel/llvm/pull/16334/commits/19983fe9bd65e3f9008cd3e244886c89dd547a3d

File: sycl/source/detail/persistent_device_code_cache.cpp (hunk 1 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/persistent_device_code_cache.cpp:323]: I am eliminating extra copy here, not intended to change types.

**Review by aelovikov-intel — COMMENTED** [review]
LGTM once the description is updated.

**Comment by KseniyaTikhomirova** [comment]
> LGTM once the description is updated.

description is updated now, thanks

**Comment by KseniyaTikhomirova** [comment]
@intel/llvm-gatekeepers  hi, this PR is ready to be merged

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 4]
```diff
@@ -990,7 +990,14 @@ ur_program_handle_t ProgramManager::getBuiltURProgram(
...
+    int Mask = 1;
+    if (URDevicesSet.size() > sizeof(Mask) * 8 - 1) {
+      // Protection for the algorithm below. Although overflow is very unlikely
+      // to be reached.
```
Comment (AlexeySachkov) [inline sycl/source/detail/program_manager/program_manager.cpp:996]: Sounds like it should be an `assert` then?

Comment (KseniyaTikhomirova) [reply sycl/source/detail/program_manager/program_manager.cpp:996]: discussed with Alexey in person, decided to make the message of exception more specific. 
done in https://github.com/intel/llvm/pull/16334/commits/637ad26731eb701dbe810dc9d2df9af1c24e8ace

**Review by aelovikov-intel — APPROVED** [review]

**Pull request merged**